cpp/ql/src/Security/CWE/CWE-020/SafeExternalAPIFunction.qll

```diff
@@ -13,9 +13,7 @@ abstract class SafeExternalAPIFunction extends Function { }
 /** The default set of "safe" external APIs. */
 private class DefaultSafeExternalAPIFunction extends SafeExternalAPIFunction {
   DefaultSafeExternalAPIFunction() {
-    // implementation note: this should be based on the properties of public interfaces, rather than accessing implementation classes directly.  When we've done that, the three classes referenced here should be made fully private.
-    this instanceof PureStrFunction or
-    this instanceof StrLenFunction or
-    this instanceof PureMemFunction
+    this instanceof ArrayFunction and
+    not this.(ArrayFunction).hasArrayOutput(_)
```

Comment:
It would be nice to have a comment explaining where this has come from, i.e. that we wish to cover string and memory buffer functions that don't have side effects.  Or would it be more accurate to say we want *any* function that doesn't have side-effects (in which case maybe the `SideEffectFunction` model could be the right tool for the job?)

Reply:
I'd also be okay with using `SideEffectFunction` in something like following way:
```codeql
exists(SideEffectFunction model | model = this |
  model.hasOnlySpecificWriteSideEffects() and
  not model.hasSpecificWriteSideEffect(_, _, _)
)
```
Given our current models, I don't think this makes any difference in the query output. For it to make a difference there would have to be a model implementation that:
- Does not implement `TaintFunction` or `DataFlowFunction`
- Implements only one of:
  + `ArrayFunction` with the current condition (i.e., without a `hasArrayOutput(_)` result).
  + `SideEffectFunction` with the condition above.

Reply:
I prefer the `SideEffectFunction` version (because I don't think the issue is really about arrays).

Reply:
Good point. I'll change it (and add a qldoc justifying the choice)

Reply:
Done in b02ac7f5232225f5bab91dce8690675dc33bebc6.

Reply:
Thanks.